lfx pre test lsp quick fix

shashank-iitbhu opened the pull request:
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

Close #1020 

#### 2. What is the scope of this PR (e.g. component or file name):

Adding LSP quick fix

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):


https://github.com/kcl-lang/kcl/assets/134289475/4cc71209-b007-44f4-904e-ac3a40b7d140



- Extending the Diagnostic Structure ```pub data: Option<serde_json::Value>```, excluded data from hash calculation due to its json type
- Used function overloading in rust to modify ```add_compile_error``` to work when data field is None
- Filled the data in ```lookup_type_from_scope()``` and ```set_type_to_scope()```
- modified ```conver_code_to_kcl_diag_id``` for CompileError
- modified ```quick_fix``` to replace new_text with suggested_replacement
- finally, extended ```kcl_diag_to_lsp_diags``` to convert data field of KCL diagnostics to data field of LSP diagnostics.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

<!-- You can choose a brief description here -->
- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

<!-- You can add more details here.
e.g. 
The test case in XXXX is used to .....
test cases in /src/tests/XXXXX
test cases https://github.com/XXX/pull/44
benchmark stats: time XXX ms
-->

He1pa commented on kclvm/error/src/diagnostic.rs:
```diff
@@ -5,11 +5,23 @@ use std::hash::Hash;
     pub level: Level,
     pub messages: Vec<Message>,
     pub code: Option<DiagnosticId>,
+    pub data: Option<serde_json::Value>,
```
you don’t need to define the data field in kcl’s diag. It is duplicated with suggested_replacement. As I explained in the step 2, the data field should be in the diag of lsp, not the diag of kcl. You can look up the function, kcl_diag_to_lsp_diags. this function convert KCL Diagnostic to LSP Diagnostics. you can see the data in result is always None. you should completed in here

shashank-iitbhu replied:
Oh right, storing the data field in KCL diagnostics is redundant. Fixed this in latest commit, and made the relevant changes in the code to utilize ```suggested_replacement``` field for conversion of KCL Diagnostic to LSP Diagnostics.

shashank-iitbhu commented:
Hey @He1pa , I have added test for the `ComplileError` Quick fix which i added before.
Now all the tests are passing. For this, I have added a compile error in `kclvm/tools/src/LSP/src/test_data/diagnostics.k` for this.

He1pa commented on kclvm/tools/src/LSP/src/quick_fix.rs:
```diff
@@ -1,17 +1,42 @@
+                    DiagnosticId::Error(error) => match error {
+                        ErrorKind::CompileError => {
+                            let replacement_text = extract_suggested_replacement(&diag.data)
+                                .unwrap_or_else(|| "".to_string());
```
if suggested resplacement is None, do nothing (just continue) instead of using an empty string.

shashank-iitbhu replied:
done.

He1pa commented on kclvm/tools/src/LSP/src/quick_fix.rs:
```diff
@@ -1,17 +1,42 @@
+                                }],
+                            );
+                            code_actions.push(CodeActionOrCommand::CodeAction(CodeAction {
+                                title: ErrorKind::CompileError.name(),
```
![image](https://github.com/kcl-lang/kcl/assets/56333845/16bf4849-aa60-4569-a51a-a6a92c6b7782)
The title is the part the user sees and should show how to fix it. For example, this is the tip given by rust-analyzer

shashank-iitbhu replied:
In the latest commit, I have provided a formatted string to the `title`:
<img width="620" alt="Screenshot 2024-02-19 at 5 01 52 PM" src="https://github.com/kcl-lang/kcl/assets/134289475/a05ca2eb-a05d-4df2-9ba2-79c0c28433ce">

Is this the way to go with this?
<br>
The rust-analyzer also provides a tip along with a clickable link to the file like this:

<img width="616" alt="Screenshot 2024-02-19 at 5 06 59 PM" src="https://github.com/kcl-lang/kcl/assets/134289475/18bda02f-3e95-41c1-8bfc-b4b49452dc02">

Do we need something like this also? If yes, how to proceed?

He1pa commented on kclvm/error/src/lib.rs:
```diff
@@ -126,14 +126,24 @@ impl Handler {
-            None,
+            suggestions,
         );
+        // println!("{:?}",suggestions.clone());
```
remove

shashank-iitbhu replied:
removed.

Peefy commented on kclvm/error/src/diagnostic.rs:
```diff
@@ -106,16 +106,16 @@ impl Diagnostic {
         range: Range,
         code: Option<DiagnosticId>,
-        suggested_replacement: Option<String>,
+        suggestions: Option<Vec<String>>,
```
s/`suggestions: Option<Vec<String>>`/`suggestions: Vec<String>`

shashank-iitbhu replied:
This will remain ```suggestions: Option<Vec<String>>```.

Peefy commented on kclvm/error/src/diagnostic.rs:
```diff
@@ -106,16 +106,16 @@ impl Diagnostic {
-                note: note.map(|s| s.to_string()),
-                suggested_replacement,
+                note: note.map(String::from),
+                suggested_replacement: suggestions.and_then(|v| v.into_iter().next()),
```
I think the suggested_replacement should be the `Vec<String>` type.

Peefy reviewed (approved):
Good Job! LGTM!
